File: python/examples/tor.py
import asyncio
from nostr_sdk import Keys, ClientBuilder, EventBuilder, Kind, KindStandard, Proxy, RelayUrl, init_logger, LogLevel, SignerAuthenticator


async def main():
    init_logger(LogLevel.INFO)

    keys = Keys.generate()
    print(keys.public_key().to_bech32())

    # Configure client to use a tor proxy for `.onion` relays
    authenticator = SignerAuthenticator(keys)
    proxy = Proxy.onion("127.0.0.1:9050")
    client = ClientBuilder().authenticator(authenticator).proxy(proxy).build()

    await client.add_relay(RelayUrl.parse("wss://relay.damus.io"))
    await client.add_relay(RelayUrl.parse("ws://oxtrdevav64z64yb7x6rjg4ntzqjhedm5b5zjqulugknhzr46ny2qbad.onion"))
    await client.add_relay(RelayUrl.parse("ws://2jsnlhfnelig5acq6iacydmzdbdmg7xwunm4xl6qwbvzacw4lwrjmlyd.onion"))
    await client.connect()

    event = EventBuilder(Kind.from_std(KindStandard.TEXT_NOTE), "Hello from rust-nostr Python bindings!").finalize(keys)
    output = await client.send_event(event)
    print("Event sent:")
    print(f" hex:    {output.id.to_hex()}")
    print(f" bech32: {output.id.to_bech32()}")
    print(f" Successfully sent to:    {output.success}")
    print(f" Failed to send to: {output.failed}")


if __name__ == '__main__':
    asyncio.run(main())

File: jvm/examples/tor.kt
package org.nostrdevkit.examples

import org.nostrdevkit.sdk.*

private class TorProxy : CustomProxy {
    override fun custom(relayUrl: RelayUrl): SocketAddr? =
        if (relayUrl.isOnion()) SocketAddr.parse("127.0.0.1:9050") else null
}

suspend fun torExample() {
    initLogger(LogLevel.INFO)

    val keys = Keys.generate()
    println(keys.publicKey().toBech32())

    val client = ClientBuilder()
        .authenticator(SignerAuthenticator(keys))
        .proxy(Proxy.custom(TorProxy()))
        .build()
    for (url in listOf(
        "wss://relay.damus.io",
        "ws://oxtrdevav64z64yb7x6rjg4ntzqjhedm5b5zjqulugknhzr46ny2qbad.onion",
        "ws://2jsnlhfnelig5acq6iacydmzdbdmg7xwunm4xl6qwbvzacw4lwrjmlyd.onion",
    )) {
        client.addRelay(RelayUrl.parse(url))
    }
    client.connect()

    val event = EventBuilder(
        Kind.fromStd(KindStandard.TEXT_NOTE),
        "Hello from rust-nostr Kotlin bindings!",
    ).finalize(keys)
    val output = client.sendEvent(event)
    println("Event sent:")
    println(" hex:    ${output.id.toHex()}")
    println(" bech32: ${output.id.toBech32()}")
    println(" Successfully sent to: ${output.success}")
    println(" Failed to send to: ${output.failed}")
}
